### DISCUSSION TITLE
What is the learning method of doing a project?

### DISCUSSION THREAD

**Discussion by NSYSUTim** [RayTracing/raytracing.github.io#1436] [Q&A]
Hope some expert can share their experience.
Hi, all.
This is my first time trying to do a project. I would like to know should I follow the instructions to the end chapter first and then comprehend the detail, modifying and adding the code, or should I follow the instructions and comprehend the detail, including the code meaning and usage, step by step at the same time?
The latter seems to be a better way of learning, but it takes a lot of time since I can learn a new concept from a single line of code, like constructor and member initialize list, include guard, member function, cerr, etc.
So I want to know what is the process of a professional doing a project.
(One more question, after we finish the instruction, should we have the ability to write all the code from scratch, or it's normal if we can't.)
Thank you.

**Comment by hollasch** [comment]
There's no single way of doing this. You can read through the books without writing any actual code at all, or you can learn and write the code as you go. Since we provide a fully-working example of this code in C++, you can just build, run and use the debugger to step through parts for greater understanding. In addition, I would never expect to "know" any material after a single pass — I'm still learning _this_ material, and I'm one of the co-authors! Personally, I've also learned to just continue grinding through material even if I don't get it. So many times I've gained insight and motivation to learn something well _after_ the time I was supposed to have understood it. My learning path has never been linear, and is usually littered with loops inside loops inside loops.

Overall, this material should be fun, and if you happen upon one approach that's more fun that some other one, I'd recommend you always pick the fun one.
